### ISSUE TITLE
Hydraulic System Monitoring by Stacking RandomForest

### ISSUE THREAD

**Issue by Gyan-17** [The-Data-Alchemists-Manipal/MindWave#266]
## 💥 Proposal

A clear and concise description of what the proposal is.

# GitHub Proposal for Predicting Hydraulic System Stability

### Dataset
[ UCI Condition monitoring of hydraulic systems Data Set](https://archive.ics.uci.edu/ml/datasets/condition+monitoring+of+hydraulic+systems)

## 1. Business Problem

The goal of this project is to predict the stability of a hydraulic system and identify the factors that cause the system to degrade. The system consists of four components: cooler condition, valve condition, internal pump leakage, and hydraulic accumulator. Each component has a degradation state that can be classified into multiple categories. The stability of the hydraulic system is a binary variable that indicates whether the system is stable or unstable.

## 2. Approach

The approach is to develop a random forest classifier for each component's degradation state using sensor data as features. Then, the predictions of the degradation states are combined with the sensor data to predict the stability of the hydraulic system using another random forest classifier. The metric used to evaluate the models is recall_weighted, which gives higher importance to the minority classes and prioritizes false negatives (the system is stable but really it is not).

## 3. Insights and Recommendations

The following insights and recommendations are derived from the feature importance analysis and the correlation matrix:

- Valve Condition: The valve condition is the most significant factor affecting the stability of the hydraulic system. When the valve condition is operating at 100%, the system is most stable. EnergyMobil should therefore focus on maintaining optimal valve conditions. Regular inspections and maintenance of the valves could prevent potential system instabilities.
- System Efficiency Sensor (SE): The system efficiency sensor (SE) is the second greatest indicator of system stability. EnergyMobil should leverage this sensor's data to predict potential system failures or instabilities. An alert system could be put in place to notify the team when the sensor's readings drop below a certain threshold, indicating a potential decrease in stability.
- Internal Pump Leakage: Internal pump leakage significantly affects the system stability. The system's stability begins to decrease when leakage starts to occur. Hence, it's crucial to monitor and mitigate any internal pump leakage promptly to ensure system stability.
- Cooler Condition: The cooler condition is not a significant factor in determining the system's stability. This can help EnergyMobil to prioritize its maintenance tasks and focus more on the valve condition and internal pump leakage.

EnergyMobil needs to focus on monitoring valve condition and internal pump leakage because once these processes drop below optimal values, the system is more likely to be unstable.

## 4. EDA

The following plots and tests are performed for exploratory data analysis:

- Histograms to check the distribution of data (independent variables)
- Boxplots to check for outliers
- Line graphs for degradation states over time
- Pairplot to show the relationship between sensor data
- Countplots for class distributions
- Correlation matrix to compare sensor and degradation states correlation
- Feature importance on all degradation states and features to find important sensors using random forest classifier
- Spearman rank test for feature importance with stable flag and sensor data (does not assume linear relationship or normally distributed data)
- Chi-squared test to see importance of conditions with stable flag

## 5. ML

The following steps are taken for machine learning:

- QuantileTransformer performs a non-linear transformation based on the ranks of the data, preserving the order of values but potentially redistributing outliers in the transformed space (want to keep outliers for failed system detection).
- SMOTE is used to handle class imbalance for degradation states (except cooler condition) and stable flag
- One hot encoder is used for degradation states for evaluating the final model

## 6. Improvements

The following improvements can be made for future work:

- Use select best features to improve performance and maintain interpretability
- Hyperparameter tuning - incorporate more hyperparameter tuning since valve condition recall score decreased by 10% when tuned. This means the model is overfitting
- Use OrdinalEncoder for the degradation conditions instead of Onehotencoder since there is an order to the classes
- Use timeseriessplit - can not use it because of the nature of the degradation conditions (specifically cooler condition) when split with different methods of timeseriessplit, folds only end up with one class or end up with very low recall scores.
- Dimensionality reduction to reduce multicollinearity between sensor data. A PCA did not separate the stable flag conditions so it was not used for model training. In the future I want to try T-SNE and UMAP methods to see if local structures and nonlinear patterns can be better captured.

I am GSSOC'23 Contributor
I am interested in doing this project
So kindly assign me this issue :)
@khusheekapoor @theyashwanthsai
